### ISSUE TITLE
Investigate why top level constructor call is not removed in Advanced mode

### ISSUE THREAD

**Issue by dimvar** [google/closure-compiler#2249]
[live demo](https://closure-compiler.appspot.com/home#code%3D%252F%252F%2520%253D%253DClosureCompiler%253D%253D%250A%252F%252F%2520%2540compilation_level%2520ADVANCED_OPTIMIZATIONS%250A%252F%252F%2520%2540output_file_name%2520default.js%250A%252F%252F%2520%253D%253D%252FClosureCompiler%253D%253D%250A%250Afunction%2520make2()%2520%257B%2520return%2520function()%2520%257B%2520return%252010%253B%2520%257D%2520%257D%250A%250Avar%2520Foo%2520%253D%2520make2()%253B%250A%250Anew%2520Foo%253B%250A)

given:
```js
// ==ClosureCompiler==
// @compilation_level ADVANCED_OPTIMIZATIONS
// @output_file_name default.js
// ==/ClosureCompiler==

function make2() { return function() { return 10; } }

var Foo = make2();

new Foo;
```

Results:
```js
new function(){return 10};
```

Expects nothing

**Comment by dimvar** [comment]
CC @mhevery

**Comment by dimvar** [comment]
Another case where we construct an object and not use it. When calling a constructor directly, the object creation gets removed, but when we add one-level of indirection, it sticks around.

```
(function() {

/** @constructor} */
function Bar() {}

var y = Bar;

new y(); // stays

new Bar(); // gets removed
 
})();
```

**Comment by concavelenz** [comment]
Removing a constructor call like everything else requires knowing that it is side-effect free, so these are all about side-effect calculations and propagation.  For the first case, I'm pretty sure it doesn't have the concept of "this is a function call that returns a side-effect free function", nor for the second case, alias propagation.  By the time it runs global constant aliases have already been eliminated (at least for the most part).

This is mostly a algorithm efficiency/compile time trade off.  Clearly it is possible to do a inference like we do for type but that is both more complex and potentially more costly time wise.

**Comment by mhevery** [comment]
> @concavelenz  This is mostly a algorithm efficiency/compile time trade off. Clearly it is possible to do a inference like we do for type but that is both more complex and potentially more costly time wise.

To me the above sounds like `won't fix` which is a reasonable response, but than I will argue that we need an escape hatch so that we can tell closure compiler that something is side-effect-free. Currently that can only be done in an extern file. I think we should have something similar for regular code as well, since as the above example demonstrates there are cases which the CC can't correctly identify as side-effect-free.

**Comment by concavelenz** [comment]
This is more of an enhancement request. It is just a question of priority of the investigation. It is possible this could be done without significantly more time by some clever ordering of work.

My main point in commenting, was this is a design level problem, not a tweak/bug fix.

**Comment by mhevery** [comment]
Thanks for the update.

Could we reopen the discussion on having an escape hatch where we could just tell CC that this has no side effects? I have opened a separate issue:  https://github.com/google/closure-compiler/issues/2256

**Comment by dimvar** [comment]
I submitted a patch yesterday to optimize the second example in this issue better. For the first example, I verified that if the side-effect calculation is moved inside the optimization loop, then we get the desired result. However, I tried that for our regression tests, and it makes very little difference, around 0.1% smaller code. And since the side-effect pass is pretty slow, it's not worth moving to the loop for so little gain. @tadeegan suggested in another thread that the accuracy of the pass itself could be improved; that's a more promising area for improvement than putting the pass in the loop.
